File: src/backend/api/git/repositories.py
"""repositories 관련 Git API 라우트."""

import logging

from fastapi import APIRouter, HTTPException

from models.git import (
    GitRepository,
    GitRepositoryCreate,
    GitRepositoryListResponse,
    GitRepositoryUpdate,
    delete_git_repository,
    get_git_repository,
    list_git_repositories,
    register_git_repository,
    update_git_repository,
)

logger = logging.getLogger(__name__)

router = APIRouter()


# =============================================================================
# Git Repository Registry Endpoints
# =============================================================================


@router.get("/repositories", response_model=GitRepositoryListResponse)
async def list_repositories():
    """List all registered Git repositories."""
    repos = list_git_repositories()
    return GitRepositoryListResponse(repositories=repos, total=len(repos))


@router.post("/repositories", response_model=GitRepository)
async def create_repository(request: GitRepositoryCreate):
    """Register a new Git repository."""
    from pathlib import Path

    from models.project import normalize_path

    # Normalize path
    path = normalize_path(request.path)

    # Validate path exists
    if not Path(path).exists():
        raise HTTPException(status_code=400, detail=f"Path does not exist: {path}")
    if not Path(path).is_dir():
        raise HTTPException(status_code=400, detail=f"Path is not a directory: {path}")

    repo = register_git_repository(
        name=request.name,
        path=path,
        description=request.description,
    )

    if not repo.is_valid:
        # Still register but warn
        logger.warning(f"Registered path '{path}' is not a valid Git repository")

    return repo


@router.get("/repositories/{repo_id}", response_model=GitRepository)
async def get_repository(repo_id: str):
    """Get a Git repository by ID."""
    repo = get_git_repository(repo_id)
    if not repo:
        raise HTTPException(status_code=404, detail="Repository not found")
    return repo


@router.put("/repositories/{repo_id}", response_model=GitRepository)
async def update_repository(repo_id: str, request: GitRepositoryUpdate):
    """Update a Git repository."""
    from pathlib import Path

    from models.project import normalize_path

    # Validate path if provided
    path = request.path
    if path:
        path = normalize_path(path)
        if not Path(path).exists():
            raise HTTPException(status_code=400, detail=f"Path does not exist: {path}")
        if not Path(path).is_dir():
            raise HTTPException(status_code=400, detail=f"Path is not a directory: {path}")

    repo = update_git_repository(
        repo_id=repo_id,
        name=request.name,
        description=request.description,
        path=path,
    )

    if not repo:
        raise HTTPException(status_code=404, detail="Repository not found")

    return repo


@router.delete("/repositories/{repo_id}")
async def remove_repository(repo_id: str):
    """Delete a Git repository from registry."""
    success = delete_git_repository(repo_id)
    if not success:
        raise HTTPException(status_code=404, detail="Repository not found")
    return {"success": True, "message": "Repository removed"}

File: src/dashboard/src/stores/git/repositories.ts
/** 등록 저장소 도메인 액션. `api/git/repositories.py` 와 대칭이다. */
import { apiClient } from '../../services/apiClient'
import type { GitRepository, GitState } from './types'

/** `orchestration/wsConnection.ts` 와 같은 형태. set/get 을 명시 인자로 받는다. */
type SetFn = (state: Partial<GitState> | ((state: GitState) => Partial<GitState>)) => void
type GetFn = () => GitState

export async function fetchRepositories(set: SetFn, _get: GetFn) {
  set({ isLoading: true, error: null })
  try {
    const data = await apiClient.get<{ repositories: GitRepository[] }>('/api/git/repositories')
    set({ repositories: data.repositories, isLoading: false })
  } catch (error) {
    set({ error: (error as Error).message, isLoading: false })
  }
}

export async function createRepository(
  set: SetFn,
  get: GetFn,
  name: string,
  path: string,
  description = ''
) {
  set({ isLoading: true, error: null })
  try {
    const repo = await apiClient.post<GitRepository>('/api/git/repositories', { name, path, description })
    await get().fetchRepositories()
    set({ isLoading: false })
    return repo
  } catch (error) {
    set({ error: (error as Error).message, isLoading: false })
    return null
  }
}

export async function updateRepository(
  set: SetFn,
  get: GetFn,
  repoId: string,
  updates: { name?: string; path?: string; description?: string }
) {
  set({ isLoading: true, error: null })
  try {
    await apiClient.put(`/api/git/repositories/${repoId}`, updates)
    await get().fetchRepositories()
    set({ isLoading: false })
    return true
  } catch (error) {
    set({ error: (error as Error).message, isLoading: false })
    return false
  }
}

export async function deleteRepository(set: SetFn, get: GetFn, repoId: string) {
  set({ isLoading: true, error: null })
  try {
    await apiClient.delete(`/api/git/repositories/${repoId}`)
    await get().fetchRepositories()
    set({ isLoading: false })
    return true
  } catch (error) {
    set({ error: (error as Error).message, isLoading: false })
    return false
  }
}
